content/posts/unpublished/ways-to-improve-be-api-performance/post.md

```diff
@@ -0,0 +1,304 @@
+  The comparison of these two approaches is shown in the diagram.
+
+![](assets/parallel.png)
+  
+  To implement the parallel approach, `CompletableFuture` may be used, CompletableFuture is a class introduced in Java 8 that allows us to write asynchronous, non-blocking code. Let's look at the example:
+  ```Java
+{
+    var productFuture=CompletableFuture.supplyAsync(()->ProductService.fetch(request),exec);
+    var paymentFuture=CompletableFuture.supplyAsync(()->PaymentService.fetch(request),exec);
+    var orderFuture=CompletableFuture.supplyAsync(()->OrderService.fetch(request),exec);
+
+    CompletableFuture<Void> allFutures=CompletableFuture.allOf(productInfoFuture,paymentFuture,orderFuture);
+
+    try {
+        allFutures.get(); // Waits until all tasks are completed
+    } catch(InterruptedException|ExecutionException e){
+        e.printStackTrace();
+    }
+
+    try {
+        requiredData.setProductInfo(productFuture.get());
+        requiredData.setPaymentInfo(paymentFuture.get());
+        requiredData.setOrderInfo(orderFuture.get());
+    } catch(InterruptedException|ExecutionException e){
+        e.printStackTrace();
+    }
+    return requiredData;
+}
+  ```
+  
+  Virtual threads are a new feature introduced in Java 21. Virtual threads are lightweight threads that reduce the effort of writing, maintaining, and debugging high-throughput concurrent applications.
+  Unlike regular thread, virtual threads does not entail any OS-level blocking. Let's try to implement the previous example using Virtual Threads.
+
+  ```Java
+{
+    Thread t1=startVirtualThread(()->requiredData.setProductInfo(ProductService.fetch(request)));
+    Thread t2=startVirtualThread(()->requiredData.setPaymentInfo(PaymentService.fetch(request)));
+    Thread t3=startVirtualThread(()->requiredData.setOrderInfo(OrderService.fetch(request)));
+
+    t1.start();t2.start();t3.start();
+    t1.join();t2.join();t3.join();
+    return requiredData;
+}
+  ```  
+
+  Spring Webclient is another tool we can use to make parallel service calls. WebClient uses an asynchronous, non-blocking solution provided by the Spring Reactive framework:
+
+  ```Java
+
+  public void run() throws Exception {
+    IntStream.range(0, 5)
+        .forEach(
+            index -> webClient.get().uri("/api/sample")
+                .retrieve()
+                .bodyToMono(String.class)
+                .publishOn(Schedulers.fromExecutor(executor))
+                .subscribe(response -> log.info("Result: {}", response))
+        );
+  }
+  ```
+
+  Using batches is another use case where parallel processing may be applied. Whether by providing a Batch API or by batch job implementation you can use `Executor Framework` or parallel stream
+for parallel processing. 
+
+  ```Java
+    public void processItems(List<Item> items) {
+      // parallel stream sample
+      items
+        .stream()
+        .parallel()
```

Comment:
good to mention that using parallel without explicitly defining an executor here might lead to problems if your task requires IO operations

Reply:
updated